File: Assignment 02/Program 07/prog7.c
//Write a C program to implement circular queue.

#include<stdio.h>
#include<stdlib.h>

//Initializing the queue structure
struct queue{
    int size;
    int f;
    int r;
    int *arr;
};

int isFull(struct queue * q){
    if((q->r+1)%q->size == q->f)
        return 1;
    return 0;
}

int isEmpty(struct queue * q){
    if(q->r == q->f)
        return 1;
    return 0;
}

//Declearing the function enqueue()
void enqueue(struct queue * q){
    int val;
    if(isFull(q)){
        printf("Queue overflow");
    }
    else{            
        printf("Enter the number :");
        scanf("%d",&val);
        q->r= (q->r+1)%q->size;
        q->arr[q->r] = val;
    }
}

//Declearing the function dequeue()
int dequeue(struct queue *q){
    int a = -1;
    if(isEmpty(q)){
        printf("Queue underflow");
    }
    else{
        q->f = (q->f+1)%q->size;
        a = q->arr[q->f];
    }
    return a;
}

//Display the array
void display(struct queue * q){
    if(isEmpty(q)){
        printf("Empty queue");
    }
    else{
        for (int i = q->f+1; i <= q->r; i++)
        {
            printf("%d ",q->arr[i]);
        }
        
    }
}
int main(){
    //Initialize and dynamic memory alloction
    struct queue * q= (struct queue*)malloc(sizeof(struct queue));
    q->f = q->r = 0;
    q->size = 8;
    q->arr = (int*)malloc(q->size*sizeof(int));
    
    //Make a chioce
    int ch;
    do{
        printf("\n____Menu____\nEnter your choice: \n1.Enqueue \n2.Dequeue \n3.Display\n4.Exit_____________________\n\n");
        scanf("%d",&ch);
        switch (ch)
        {
        case 1:
            enqueue(q);
            break;
        case 2:
            printf("Dequed item : %d\n",dequeue(q));
            break;
        case 3:
            display(q);
            break;
        case 4:
            printf("Exiting...");
            break;
        default:
            printf("You give wrong input");
            break;
        }
    }while(ch!= 4);
    free(q->arr);
    return 0;
}

File: Assignment 02/Program 07/prog7.py
# Write a Python program to implement circular queue using array.

class queue:
    def __init__(self,size):
        self.size = size
        self.front = -1
        self.rear = -1
        self.queue = [None]*size

    def enqueue(self,item):
        if (self.rear+1)%self.size == self.front:
            print("Queue Overflow")
            return
        if self.front == -1:  
            self.front = 0
            self.rear = 0
        else:
            self.rear = (self.rear+1)%self.size
        self.queue[self.rear] = item

    def dequeue (self):
        if self.front== -1:
            print("Queue Underflow")
            return
        data = self.queue[self.front]
        if self.front == self.rear:
            self.front = -1
            self.rear = -1
        else:
            self.front = (self.front+1)%self.size
        return data
    
    def display (self):
        if self.front == -1:
            print("Empty Queue")
        else:
            i = self.front
            while True:
                print(self.queue[i],end= " ")
                if i == self.rear:
                    break
                i = (i+1)%self.size
      

q = queue(5)

n=0
while n!= 4:
    n = int(input("Enter the choice:\n1.Enqueue\n2.Dequeue\n3.Display\n4.Exit\n\n"))
    match n:
        case 1:
            val = int(input("Enter the number: "))
            q.enqueue(val)
        case 2:
            print("Poped element:",q.dequeue())
        case 3:
            q.display()
    
        
